### PR TITLE
Detailed device view

### LINKED ISSUE #14: Create detailed device view
Create detailed device view

### LINKED ISSUE #21: Create detailed device controller
Create controller to do remote commands

### LINKED ISSUE #22: Create detailed device widget controller
Create controller to render a detailed device

### PR DESCRIPTION
**Pull request by PatricioSaldivar**
Added detailed device view.

### CODE REVIEW

File: lib/controllers/widget/custom_fields_card.dart [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+                                  style: TextStyle(
+                                      fontSize: 18,
+                                      color: Colors.black87,
+                                      fontWeight: FontWeight.w500),
+                                  overflow: TextOverflow.clip,
+                                ),
+                                Text(
+                                  widget._device.annotatedAssetId == null
+                                      ? '-'
+                                      : widget._device.annotatedAssetId,
```
Comment (Diego1149) [inline lib/controllers/widget/custom_fields_card.dart]: For all of these instead of using the same logic use a function getValueOrDefault(widget._device.property)

File: lib/controllers/widget/detailed_device.dart [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+      setState(() {
+        _loading = false;
+        _errorOnLoading = true;
+        _warning = e;
+      });
+    });
+  }
+
+  /// Pops an alert to user and sends the app to the login view.
+  _alertAndLogIn(ErrorHandler error) {
```
Comment (Diego1149) [inline lib/controllers/widget/detailed_device.dart]: All of these alert functions shouldn't be here, they should be in a centric point and shared between all code. Maybe you can put them in the ErrorHandler

File: lib/controllers/widget/hardware_os_card.dart [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+                              Text(
+                                'Last Enrollment Time',
+                                style: TextStyle(
+                                    fontSize: 18,
+                                    color: Colors.black87,
+                                    fontWeight: FontWeight.w500),
+                                overflow: TextOverflow.clip,
+                              ),
+                              Text(
+                                widget._device.lastEnrollmentTime,
```
Comment (Diego1149) [inline lib/controllers/widget/hardware_os_card.dart]: Except for like 3-5 properties everything is optional, so can you please wrap all of the properties around the function getValueOrDefault

Comment (Diego1149) [reply lib/controllers/widget/hardware_os_card.dart]: All the ones for which it applies, because for example in auto update expiration it doesn't apply.

File: lib/controllers/widget/detailed_device.dart (hunk 2 above)
Comment (PatricioSaldivar) [reply lib/controllers/widget/detailed_device.dart]: I already move them to Error Handler class, thanks for the advice.

File: lib/controllers/widget/hardware_os_card.dart (hunk 3 above)
Comment (PatricioSaldivar) [reply lib/controllers/widget/hardware_os_card.dart]: Done.

File: lib/controllers/widget/custom_fields_card.dart (hunk 1 above)
Comment (PatricioSaldivar) [reply lib/controllers/widget/custom_fields_card.dart]: Done.

**Review by abbiePetch — APPROVED** [review]

**Comment by abbiePetch** [comment]
Many UI config looks repeat. Please try to refactor hardware and custom fields cards

**Comment by abbiePetch** [comment]
Also refactor all UI config as much as you can. It's very long and hard to follow. One idea is to add UI/Common folder and put all common ui widget there (ie headers, body, tile, menu).

**Comment by PatricioSaldivar** [comment]
> Also refactor all UI config as much as you can. It's very long and hard to follow. One idea is to add UI/Common folder and put all common ui widget there (ie headers, body, tile, menu).

I already added a issue for that #47, I will resolve it on the next PR.

File: lib/controllers/widget/custom_fields_card.dart [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+                                Text(
+                                  'User',
+                                  style: TextStyle(
+                                      fontSize: 18,
+                                      color: Colors.black87,
+                                      fontWeight: FontWeight.w500),
+                                  overflow: TextOverflow.clip,
+                                ),
+                                Text(
+                                  widget._device.annotatedUser == null
```
Comment (pinkfish) [inline lib/controllers/widget/custom_fields_card.dart]: You can do this FYI.  widget._device.annotedUser ?? "-"

Will return the "-" only if the first piece is null.

File: lib/controllers/widget/custom_fields_card.dart [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+                              )
+                            ],
+                          ),
+                        ),
+                      ],
+                    )),
+              ],
+            ),
+            Padding(padding: EdgeInsets.all(0), child: Divider()),
+            Row(
```
Comment (pinkfish) [inline lib/controllers/widget/custom_fields_card.dart:148]: Why not make this a ButtonBar and FlatButton?  Rather than an inkwell?

File: lib/controllers/widget/custom_fields_card.dart (hunk 4 above)
Comment (PatricioSaldivar) [reply lib/controllers/widget/custom_fields_card.dart]: That is really good to know, thank you.

File: lib/controllers/widget/custom_fields_card.dart (hunk 5 above)
Comment (PatricioSaldivar) [reply lib/controllers/widget/custom_fields_card.dart:148]: Great idea, done.

File: lib/controllers/widget/custom_fields_card.dart (hunk 1 above)
Comment (pinkfish) [reply lib/controllers/widget/custom_fields_card.dart]: Actually, should do this using the ?? operator.  ie: widget._device.annotatedAssertId ?? '-'  don't use a function ;)

Comment (PatricioSaldivar) [reply lib/controllers/widget/custom_fields_card.dart]: Got it :)

File: lib/controllers/widget/hardware_os_card.dart [hunk 6]
```diff
@@ -0,0 +1,405 @@
...
+                      Padding(
+                          padding: EdgeInsets.only(top: 10, bottom: 10),
+                          child: Column(
+                            crossAxisAlignment: CrossAxisAlignment.start,
+                            children: <Widget>[
+                              Text(
+                                'OS version',
+                                style: TextStyle(
+                                    fontSize: 18,
+                                    color: Colors.black87,
```
Comment (Diego1149) [inline lib/controllers/widget/hardware_os_card.dart:292]: Since most of these TextStyles are repeated can we have a variable that has it and we use that instead?

File: lib/controllers/widget/hardware_os_card.dart [hunk 7]
```diff
@@ -0,0 +1,405 @@
...
+                                'Device ID',
+                                style: TextStyle(
+                                    fontSize: 18,
+                                    color: Colors.black87,
+                                    fontWeight: FontWeight.w500),
+                                overflow: TextOverflow.clip,
+                              ),
+                              Text(
+                                widget._device.deviceId ?? '-',
+                                style: TextStyle(
```
Comment (Diego1149) [inline lib/controllers/widget/hardware_os_card.dart:340]: Same for this one

**Review by Diego1149 — APPROVED** [review]

File: lib/controllers/widget/hardware_os_card.dart (hunk 6 above)
Comment (PatricioSaldivar) [reply lib/controllers/widget/hardware_os_card.dart:292]: I refactored the UI in another pull request.

File: lib/controllers/widget/hardware_os_card.dart (hunk 7 above)
Comment (PatricioSaldivar) [reply lib/controllers/widget/hardware_os_card.dart:340]: I refactored the UI in another pull request.
